extensions/smallrye-reactive-messaging/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/deployment/SmallRyeReactiveMessagingProcessor.java

```diff
@@ -143,54 +149,98 @@ void validateBeanDeployment(
+                            .stream()
+                            .filter(ai -> io.quarkus.smallrye.reactivemessaging.deployment.DotNames.LEGACY_ON_OVERFLOW
+                                    .equals(ai.name()))
+                            .filter(ai -> {
+                                if (ai.target().kind() == AnnotationTarget.Kind.METHOD_PARAMETER && injectionPoint
+                                        .isParam()) {
+                                    return ai.target().asMethodParameter().position() == injectionPoint.getPosition();
+                                }
+                                return true;
+                            })
+                            .findAny();
+                    createEmitter(emitters, injectionPoint, channelName, overflow);
                 }
             }
         }
     }
 
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    private void createEmitter(BuildProducer<EmitterBuildItem> emitters, InjectionPointInfo injectionPoint,
+            String channelName,
+            Optional<AnnotationInstance> overflow) {
+        LOGGER.debugf("Emitter injection point '%s' detected, channel name: '%s'",
+                injectionPoint.getTargetInfo(), channelName);
+        if (overflow.isPresent()) {
+            AnnotationInstance annotation = overflow.get();
+            AnnotationValue maybeBufferSize = annotation.value("bufferSize");
+            int bufferSize = maybeBufferSize != null ? maybeBufferSize.asInt() : 0;
+            emitters.produce(
+                    EmitterBuildItem.of(channelName, annotation.value().asString(), bufferSize));
+        } else {
+            emitters.produce(EmitterBuildItem.of(channelName));
+        }
+    }
+
     @BuildStep
     public List<UnremovableBeanBuildItem> removalExclusions() {
         return Arrays.asList(
                 new UnremovableBeanBuildItem(
-                        new BeanClassAnnotationExclusion(io.quarkus.smallrye.reactivemessaging.deployment.DotNames.INCOMING)),
+                        new BeanClassAnnotationExclusion(
+                                io.quarkus.smallrye.reactivemessaging.deployment.DotNames.INCOMING)),
                 new UnremovableBeanBuildItem(
-                        new BeanClassAnnotationExclusion(io.quarkus.smallrye.reactivemessaging.deployment.DotNames.OUTGOING)));
+                        new BeanClassAnnotationExclusion(
+                                io.quarkus.smallrye.reactivemessaging.deployment.DotNames.OUTGOING)));
+    }
+
+    @BuildStep
+    public void enableMetrics(Capabilities capabilities, ReactiveMessagingConfiguration configuration,
+            BuildProducer<AdditionalBeanBuildItem> beans) {
+        if (capabilities.isCapabilityPresent(Capabilities.METRICS) && configuration.metricsEnabled) {
+            beans.produce(new AdditionalBeanBuildItem(MetricDecorator.class.getName()));
```

Comment:
I'm trying it and it looks like the `MetricDecorator` is included always (!), even without this line for some reason. And if the Metrics dependency is not present on the application, then it fails to boot due to a `ClassNotFoundException: org.eclipse.microprofile.metrics.MetricRegistry` because the `MetricDecorator` references that class. We need to find out why this bean is being included even when we don't want it.

Reply:
Hm, I think it's because SR-RM provider jar contains a `beans.xml` so all its beans are automatically picked up: 
https://github.com/smallrye/smallrye-reactive-messaging/blob/master/smallrye-reactive-messaging-provider/src/main/resources/META-INF/beans.xml
Not sure if we can specifically exclude `MetricDecorator` without having to get rid of that `beans.xml`

Reply:
It contains a bean.xml as it was requested. But yes, I can see the issue... Not sure how I can work around this issue. Dropping the beans.xml in reactive messaging would be seen as a breaking change.

Reply:
What's the purpose of it being there? What's the use case it was needed for?

Reply:
We could potentially "remove" the `@ApplicationScoped` annotation from that bean by an annotation transformer, but that's ugly. Or add a build-time mechanism for Arc for excluding a particular bean, if there's no such thing yet. Also quite ugly.
I feel we should get rid of the `beans.xml` if possible. If it's needed during testing for example, it can be added using ShrinkWrap. Removing it could be a breaking change from the PoV of the runtime that imports SR-RM, but if implemented properly it shouldn't be of any concern to user applications, I guess

Reply:
Removing the `beans.xml` is a dead end. It breaks too many things (such as the Java SE support). 

One approach would be to move the metric decorator to its own module - without a beans.xml, and add the bean only in Quarkus. That potentially means more work to pass the TCK (mandating the metric support) and providing an implementation that would only be compliant in Quarkus. So not sure about this approach either.

Reply:
One more option would be to make the decorator a no-op when the metrics extension is not available. To not trigger a `ClassNotFoundException` we would have to get rid of the `MetricRegistry` injection in it, and instead retrieve the registry by using static methods of the SmallRye-specific `MetricRegistries` class.
Not that it's nice, but it's possible.

Reply:
That would require a change in SR Reactive Messaging that would use an SR specific API? Unfortunately, it's not possible as this implementation is used in other environments (that do not use SR Metrics)

Reply:
I discussed with @mkouba and it seems that using a transformer to veto the bean is the best approach.

Reply:
Hm, yes that would be using SmallRye-specific API.
Vetoing sounds like a good idea!

Reply:
I've implemented the veto approach.